### ISSUE TITLE
Cilium 1.11.0 does not start on EKS with pod subnet-filter

### ISSUE THREAD

**Issue by mhulscher** [cilium/cilium#18239]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I am installing Cilium 1.11.0 on a brand new EKS cluster on nodes that did not previously have a CNI installed. The cilium-agents do not become ready. They create CiliumNode resources. However, the cilium-operator does not allocate ENIs/IPs to the nodes. As a result the nodes don't become ready. The VPCs primary and secondary CIDRs don't seem to be discovered either.

### Cilium Version

1.11.0

### Kernel Version

5.4.156-83.273.amzn2.x86_64

### Kubernetes Version

v1.21.2-eks-06eac09

### Sysdump

I don't feel comfortable sharing this in a public github issue. Please reach out and I can share this privately.

### Anything else?

Installing cilium 1.10.5 with the same values.yaml works well.

EDIT: Just tested, the same values also work with 1.10.6. So the problem is unlikely to be present in any of the fixes backported to 1.10

Collected log files:

[cilium-1-11-0.zip](https://github.com/cilium/cilium/files/7696918/cilium-0-11-0.zip)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by mhulscher** [comment]
Interestingly enough I can upgrade Cilium on existing nodes from 0.10 to 0.11. However, new nodes don't become ready with 0.11.

**Comment by pchaigno** [comment]
> I am installing Cilium 0.11.0

I guess you meant Cilium 1.11.0? We've had stable releases for years now :sweat_smile: 

In the cilium-operator logs, I'm seeing a lot of warnings for different nodes:
```
level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-02ad8b4c55e4a7c08 name=ip-10-28-92-98.eu-west-1.compute.internal subsys=ipam
```
It's probably because the operator can't find the nodes that it is not allocating ENIs and IP addresses.

Could you please share the `ciliumnodes-xxx.yaml` and `k8s-nodes-xxx.yaml` files from the sysdump?

**Comment by mhulscher** [comment]
Yes, this relates to 1.11.0 :') I will reproduce the issue again tomorrow and attach the ciliumnodes and k8s-nodes

**Comment by mhulscher** [comment]
I've attached the files.
[cilium-sysdump-20211215-113339.zip](https://github.com/cilium/cilium/files/7718565/cilium-sysdump-20211215-113339.zip)

**Comment by azharprabudi** [comment]
i face the same problem when configure filter subnets by id or tags

**Comment by pchaigno** [comment]
@mhulscher @azharprabudi Could you share the exact reproduction steps if you have them?

**Comment by amilanoski** [comment]
I am facing this issue in 1.10.6

eni portion helm values.yaml
```
eni:
  # -- Enable Elastic Network Interface (ENI) integration.
  enabled: true
  # -- Update ENI Adapter limits from the EC2 API
  updateEC2AdapterLimitViaAPI: true
  # -- Release IPs not used from the ENI
  awsReleaseExcessIPs: true
  # -- EC2 API endpoint to use
  ec2APIEndpoint: ""
  # -- Tags to apply to the newly created ENIs
  eniTags: {}
  # -- If using IAM role for Service Accounts will not try to
  # inject identity values from cilium-aws kubernetes secret.
  # Adds annotation to service account if managed by Helm.
  # See https://github.com/aws/amazon-eks-pod-identity-webhook
  iamRole: "arn:aws:iam::xxxx:role/cilium_operator_serviceaccount_role"
  # -- Filter via subnet IDs which will dictate which subnets are going to be used to create new ENIs
  # subnetIDsFilter: ""
  subnetIDsFilter: "subnet-097xxxxxfabc3,subnet-0f7cbxxxa4d3d,subnet-0892xxx003d7"
  # -- Filter via tags (k=v) which will dictate which subnets are going to be used to create new ENIs
  # subnetTagsFilter: function=eks
  subnetTagsFilter: ""
```

Logs: in operator showing similar logs

```
level=info msg="Synchronized ENI information" numInstances=0 numSecurityGroups=12 numSubnets=0 numVPCs=1 subsys=eni
level=debug msg="Resetting resyncNeeded" instanceID=i-07a87xxxa424d1 name=ip-100-88-xx-x.ec2.internal subsys=ipam
level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-07a87xxx7aa424d1 name=ip-100-88-xx-xx.ec2.internal subsys=ipam
level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-07a87xxxa424d1 name=ip-100-88-xx-xx.ec2.internal subsys=ipam
level=debug msg="Resetting resyncNeeded" instanceID=i-00980xxx4890cca name=ip-100-88-xx-xx.ec2.internal subsys=ipam
```

however when I update value **subnetTagsFilter** with like below

`subnetTagsFilter: function=eks`

logs show
`level=info msg="Synchronized ENI information" numInstances=11 numSecurityGroups=12 numSubnets=11 numVPCs=1 subsys=eni`

Notice the **numSubnets**

It looks like the operator is filtering correctly based on the parameters that are being passed to it.

**Comment by amilanoski** [comment]
Upgraded to 1.11.0
I was able to get the subnetFilter to work, but then the operator could not get the security group.

Notice numSubnet is 3 now.
`level=info msg="Synchronized ENI information" numInstances=4 numSecurityGroups=12 numSubnets=3 numVPCs=1 subsys=eni`

`level=warning msg="Unable to create interface on instance: unable to get security groups failed to get security group ids" instanceID=i-00d45xxxx06d4c name=ip-100-88-xx-xx.ec2.internal subsys=ipam`

**Comment by gandro** [comment]
Might be relevant here: We have discovered a bug in way `subnet-tags-filter` is parsed, see #18293 

The TLDR is that `subnetTagsFilter: "foo=bar"` does not work when provided via ConfigMap, but you may use a JSON filter as a workaround, i.e. `subnetTagsFilter: '{"foo":"bar"}'`. We plan to fix that.

But the above bug has been present since the introduction of the `subnet-tags-filter` flag was added. We have confirmed that the CLI parse bug is alo present in v1.10, therefore I don't understand why some report that the flag works for them in v1.10 - have you been using the JSON syntax all along?

**Comment by gandro** [comment]
> Hi @gandro ,
> 
> What works for me in 1.10 is to set `eni.subnetTagsFilter: "foo=bar"` in the `values.yaml`. However, I don't set the filter in the CNI ConfigMap.

Which v1.10 version are you using? My understanding is that the `eni.subnetTagsFilter` will be silently ignored in such a setup due the above bug. But we might be missing something else here.

You can verify that the really filter works by checking that the `numSubnets` count for the `level=info msg="Synchronized ENI information"` message in the cilium-operator? It should report the same number of subnets as `aws ec2 describe-subnets --filters 'Name=tag:foo,Values=bar'` returns.

**Comment by mhulscher** [comment]
So I am installing cilium 1.10.6, but this has also worked with 1.10.5.

Given a `cilium-config` ConfigMap with 
```
  subnet-tags-filter: foo=bar
```

I can see the following in the cilium-operator logs:
```
cilium-operator-6df4bbf59c-drkvb cilium-operator level=info msg="  --subnet-tags-filter='foo=bar'" subsys=cilium-operator-aws
```

> You can verify that the really filter works by checking that the numSubnets count for the level=info msg="Synchronized ENI information" message in the cilium-operator? It should report the same number of subnets as aws ec2 describe-subnets --filters 'Name=tag:foo,Values=bar' returns.

Interestingly, the operator detects 9 subnets.
> 3x public
> 3x private
> 3x pod

```
cilium-operator-6df4bbf59c-drkvb cilium-operator level=info msg="Synchronized ENI information" numInstances=36 numSecurityGroups=10 numSubnets=9 numVPCs=1 subsys=eni
```

However, I expect the operator to only detect the 3 pod-subnets:
```
$ aws ec2 describe-subnets --filters=Name=tag:foo,Values=bar | jq '.Subnets | length'
3
```

The reason why everything still works, is probably because of what you suggested; the subnets that I am targeting have way more free IPs than the other subnets.

Now when I set the following, still on 1.10.6, I see `numSubnets=3` 🎉 

```
    eni:
      subnetTagsFilter: '{"foo":"bar"}
```

When I now update cilium-operator to 1.11.0, I see the following in its output:

```
level=fatal msg="Incorrect config-map flag subnet-tags-filter" error="{\"foo\":\"bar\"} must be formatted as key=value" subsys=config
```

Now on 1.11.0, I set `  subnet-tags-filter: foo=bar` and I can see `numSubnets=3`, still correct.

However, new nodes are not initialized properly by the operator. The operator logs the following when a new node is created:
```
cilium-operator-7857545c66-ckqgc cilium-operator level=info msg="Discovered new CiliumNode custom resource" name=ip-10-28-83-186.eu-west-1.compute.internal subsys=ipam
cilium-operator-7857545c66-ckqgc cilium-operator level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-0045e33f849ad944b name=ip-10-28-83-186.eu-west-1.compute.internal subsys=ipam
cilium-operator-7857545c66-ckqgc cilium-operator level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-0045e33f849ad944b name=ip-10-28-83-186.eu-west-1.compute.internal subsys=ipam
```

I think this is happening here: https://github.com/cilium/cilium/blob/master/pkg/aws/eni/node.go#L478

What could perhaps be going wrong, is that I am also using `first-index-interface: 1`, like so in the cni-configuration configmap:

```
{
  "cniVersion": "0.3.1",
  "name": "cilium",
  "plugins": [
    {
      "cniVersion": "0.3.1",
      "eni": {
        "first-interface-index": 1
      },
      "type": "cilium-cni"
    }
  ]
}
```

The first interface lives inside the private subnet, which does not match the subnet-tags-filter. There are no other ENIs attached; we expect cilium to do that. The ENIs that we expect cilium to create should be placed in the pod subnets, which do match the subnet-tags-filter.

**Comment by gandro** [comment]
Thanks for the information, that's helpful. I wasn't aware that the `k=v` syntax throws an error on v1.11. 

It seems like we might have two bugs here:
- One is the parser issue (which doesn't seem present on v1.11)
- The other one is the `instance not found` error, for which we have not yet established a root cause

> ```
> cilium-operator-7857545c66-ckqgc cilium-operator level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-0045e33f849ad944b name=ip-10-28-83-186.eu-west-1.compute.internal subsys=ipam
> ```
Does this instance (`ip-10-28-83-186.eu-west-1.compute.internal`) actually exist in your cluster? Or is this error indeed caused by a stale CiliumNode resource?

> The first interface lives inside the private subnet, which does not match the subnet-tags-filter. 
Yes, that is okay. The subnet-tag-filter only applies to ENIs created/managed by Cilium.

> There are no other ENIs attached; we expect cilium to do that. The ENIs that we expect cilium to create should be placed in the pod subnets, which do match the subnet-tags-filter.
Yes, I agree that this should be the expected behavior.

**Comment by mhulscher** [comment]
> Does this instance (ip-10-28-83-186.eu-west-1.compute.internal) actually exist in your cluster? Or is this error indeed caused by a stale CiliumNode resource?

The node actually exists. The operator is throwing these errors because it is triggered by a new CiliumNode, which is created by the cilium-agent running on the newly created k8s node. The error that the node can't be found is thrown when cilium can't discover any ENIs for the node (as per the function that I linked).

**Comment by mhulscher** [comment]
Perhaps the issue I am having is case a of PEBKAC, maybe I am misunderstanding the documentation?

My goal is the as follows:

- Nodes are created with their first interface in a private subnet
- Cilium should discover all subnets in the VPC; it should be aware of all CIDRs in the VPC so that traffic to other subnets is not SNATted
- Cilium should assign ENIs and IPs from the pod subnets to pods
- Traffic with destination **inside** the VPC should not be SNATted
- Traffic with destination **outside** the VPC **should** be SNATted using the node's first interface's primary IP as source IP (I have set `bpf.masquerade: true`).

So far, with 1.10.x, I have only set `eni.subnetTagsFilter` and I did not set the `eni.subnet-tags` in the CNI configuration. We've established that this configuration does not work with 1.11.0.

However, on 1.11.0, if I **do** set `eni.subnet-tags` in the CNI configuration and **do not** set `eni.subnetTagsFilter`, everything seems to work like I expect it to.

Did I misconfigure Cilium all along?

**Comment by gandro** [comment]
> So far, with 1.10.x, I have only set eni.subnetTagsFilter and I did not set the eni.subnet-tags in the CNI configuration. We've established that this configuration does not work with 1.11.0.
> However, on 1.11.0, if I do set eni.subnet-tags in the CNI configuration and do not set eni.subnetTagsFilter, everything seems to work like I expect it to.

Given this information, I do believe might we have a bug around `eni.subnetTagsFilter` (i.e. the operator `--subnet-tags-filter` flag) in v1.11 besides the CLI parsing bug already discovered. Because I expect `eni.subnetTagsFilter` (Helm) and `eni.subnet-tags` (CNI Config) to be functionally *more or less* identical. 

In other words, both should work, and both filters should be applied (i.e. attached ENIs should have tags matching _both_ filters).

The CNI field has existed for a while, and at least to my knowledge has been used by one large Cilium user using it in production for a while. It is usually what I recommend and what is documented, because it is very flexible (as it can be configured per node).

The global Helm flag filter was added later by an external contributor for what seems to be performance reasons (#10738). I don't think it has had much usage, especially given the fact that bug #18328 has not been uncovered until recently, even though it has been present since the beginning.

TL;DR: I don't think this is just a configuration issue. However, I do believe that the `eni.subnet-tags` CNI configuration field is likely not affected by the bug reported here.

**Comment by espringsteen** [comment]
@gandro Has any progress been made on the second bug described here: 
```
The other one is the instance not found error, for which we have not yet established a root cause
```

**Comment by gandro** [comment]
Not from my side. But I'm also not actively working on this at the moment.

Given that have merged the fix for #18328 on master, I think the next step here is to try to reproduce the issue on `master` to get a clear understanding what the outstanding issues are.

**Comment by sayboras** [comment]
:wave: let me give it a try and inform the result in this issue.

**Comment by sayboras** [comment]
I have performed some test in both master and v1.10.7 releases, please find below details.

In short, while there is a warning in operator logs on instance not found, cilium agent on new node is still managed to start shortly after that, this is for both master build and v1.10.7. 


Feel free to let me know if there is anything that I could have missed here.

<details>
<summary>Installation configuration</summary>

```
$ cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: cni-configuration
  namespace: kube-system
data:
  cni-config: |-
    {
      "cniVersion":"0.3.1",
      "name":"cilium",
      "plugins": [
        {
          "cniVersion":"0.3.1",
          "type":"cilium-cni",
          "eni": {
            "subnet-tags":{
              "eksctl.cluster.k8s.io/v1alpha1/cluster-name":"tammach-17899"
            },
            "first-interface-index": 1
          }
        }
      ]
    }
EOF

$ cat aws-eni-valus.yaml
operator:
  extraArgs:
    - --aws-release-excess-ips

eni:
  enabled: true
  subnetTagsFilter: eksctl.cluster.k8s.io/v1alpha1/cluster-name=tammach-17899
  
ipam:
  mode: eni

egressMasqueradeInterfaces: eth1
tunnel: disabled

cni:
  customConf: true
  configMap: cni-configuration
  subnetTagsFilter: eksctl.cluster.k8s.io/v1alpha1/cluster-name=tammach-17899

bpf:
  masquerade: true
```

</details>


<details>
<summary>Instance addition/removal is handled gracefully in master build</summary>

```
2022-02-11T12:29:16.191827362Z level=info msg="Synchronized ENI information" numInstances=2 numSecurityGroups=5 numSubnets=3 numVPCs=2 subsys=eni
2022-02-11T12:30:16.669623152Z level=info msg="Synchronized ENI information" numInstances=1 numSecurityGroups=5 numSubnets=3 numVPCs=2 subsys=eni
2022-02-11T12:31:17.120613422Z level=info msg="Synchronized ENI information" numInstances=2 numSecurityGroups=5 numSubnets=3 numVPCs=2 subsys=eni
2022-02-11T12:32:06.163236979Z level=info msg="Discovered new CiliumNode custom resource" name=ip-192-168-116-146.ap-southeast-2.compute.internal subsys=ipam

```

</details>


<details>
<summary>Instance deletion warning, but shortly gracefully handled</summary>

```
2022-02-11T12:59:13.729822732Z level=info msg="Synchronized ENI information" numInstances=1 numSecurityGroups=5 numSubnets=9 numVPCs=2 subsys=eni
2022-02-11T12:59:13.730185570Z level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-0bdf82bd61d3c3008 name=ip-192-168-160-163.ap-southeast-2.compute.internal subsys=ipam
2022-02-11T12:59:13.730199828Z level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-0bdf82bd61d3c3008 name=ip-192-168-160-163.ap-southeast-2.compute.internal subsys=ipam
2022-02-11T12:59:13.774013593Z level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-0bdf82bd61d3c3008 name=ip-192-168-160-163.ap-southeast-2.compute.internal subsys=ipam
2022-02-11T12:59:13.774046199Z level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-0bdf82bd61d3c3008 name=ip-192-168-160-163.ap-southeast-2.compute.internal subsys=ipam
2022-02-11T13:00:14.301098157Z level=info msg="Synchronized ENI information" numInstances=2 numSecurityGroups=5 numSubnets=9 numVPCs=2 subsys=eni
2022-02-11T13:01:02.428834858Z level=info msg="Discovered new CiliumNode custom resource" name=ip-192-168-151-204.ap-southeast-2.compute.internal subsys=ipam
2022-02-11T13:01:02.428877188Z level=info msg="Resolving IP deficit of node" available=0 availableForAllocation=0 availableInterfaces=2 instanceID=i-0f4062e9b1fb60b1e maxIPsToAllocate=8 name=ip-192-168-151-204.ap-southeast-2.compute.internal neededIPs=8 remainingInterfaces=2 selectedInterface= selectedPoolID= subsys=ipam used=0


$ cilium status      
    /¯¯\
 /¯¯\__/¯¯\    Cilium:         OK
 \__/¯¯\__/    Operator:       OK
 /¯¯\__/¯¯\    Hubble:         disabled
 \__/¯¯\__/    ClusterMesh:    disabled
    \__/

DaemonSet         cilium             Desired: 2, Ready: 2/2, Available: 2/2
Deployment        cilium-operator    Desired: 2, Ready: 2/2, Available: 2/2
Containers:       cilium             Running: 2
                  cilium-operator    Running: 2
Cluster Pods:     2/2 managed by Cilium
Image versions    cilium             quay.io/cilium/cilium:v1.10.7@sha256:e23f55e80e1988db083397987a89967aa204ad6fc32da243b9160fbcea29b0ca: 2
                  cilium-operator    quay.io/cilium/operator-aws:v1.10.7@sha256:97b378e0e3b6b5ade6ae1706024c7a25fe6fc48e00102b65a6b7ac51d6327f40: 2

```

</details>

**Comment by gandro** [comment]
We also tested the above configuration without the `cni` section and it also works with the master branch. We expect the upcoming Cilium v1.11.2 to fix the issues reported here.

**Comment by prune998** [comment]
I have the exact same issue here.
Everything is working on 1.10.5, and I have mixed results with 1.11.1 or 1.11.2 (that I just started to test).

cilium-operator keeps saying:
```
level=info msg="Synchronized ENI information" numInstances=12 numSecurityGroups=24 numSubnets=3 numVPCs=1 subsys=eni
level=info msg="Resolving IP deficit of node" available=28 availableForAllocation=0 availableInterfaces=2 instanceID=i-0c5a3a0fa068db10f maxIPsToAllocate=3 name=ip-10-253-2-252.us-west-2.compute.internal neededIPs=3 remainingInterfaces=2 selectedInterface= selectedPoolID= subsys=ipam used=23
level=warning msg="Unable to create interface on instance: unable to get security groups failed to get security group ids" instanceID=i-0c5a3a0fa068db10f name=ip-10-253-2-252.us-west-2.compute.internal subsys=ipam
level=warning msg="Unable to maintain ip pool of node" error="unable to get security groups failed to get security group ids" instanceID=i-0c5a3a0fa068db10f name=ip-10-253-2-252.us-west-2.compute.internal subsys=ipam
```

So it's `unable to get security groups failed to get security group ids` and it's a `warning` only... but I have those logs many times per second.

Cilium is working with the current IPs it have defined in the `ciliumnode` but is not able to grab new / or is just not able o add new interfaces (ENIs) on the nodes...

Instance type is `m5.xlarge` which should allow 4 ENIs. I have 3 configured, the `eth0` is used only by the Node IP (10.253.x.y), then `eth1` and `eth2` used for pod ip's. No `eth3` created.
I do have 14 IPs on each 1 and 2 ENIs, which means they are full.

So why isn't the Operator able to attach a 4th ENI ? 

All this testing was done with only one node.
Then I tried adding two other nodes, and I get the same errors about `Instance not found`.
It seems to be down to the Operator not being able to talk to the AWS API (or get the needed informations from the API)

The CiliumAgent is logging:

```
level=info msg="API request released by rate limiter" name=endpoint-delete parallelRequests=38 subsys=rate uuid=03642edb-853b-47e8-975c-68fc86f2b37a waitDurationTotal="83.892µs"
level=info msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="2.839µs" subsys=rate totalDuration="105.912µs" uuid=03642edb-853b-47e8-975c-68fc86f2b37a waitDurationTotal="83.892µs"
```

So I tried to `curl http://169.254.169.254/latest/meta-data/instance-id` from a node: OK
and `curl http://169.254.169.254/latest/meta-data/instance-id` from inside a (working) POD: no answer !

Looking at Hubble, it seems the TCP cnx is `forwarded` to `169.254.169.254` (the IMDS Metadata server) 
Looking at a tcpdump on the node (eth0) I see:
```
17:48:01.537330 IP 100.65.93.117.34488 > 169.254.169.254.http: Flags [S], seq 1318068199, win 62727, options [mss 8961,sackOK,TS val 2677816022 ecr 0,nop,wscale 7], length 0
```

So the traffic is clearly not flowing.
Again, this is totally working fine with Cilium 1.10.x

**Comment by prune998** [comment]
I'm using the CNI config as:
```
  cni-config: |-
    {
      cniVersion":"0.3.1",
      "name": cilium",
      plugins": [
        {
          cniVersion": "0.3.1",
          type": "cilium-cni",
          eni": {
            first-interface-index": 1,
            subnet-tags":{
              "Usage":"pods"
            }
          }
        }
      ]
    }
```

Note that I also enabled the `ip-masq-agent` :

``` k   exec -it -n kube-system cilium-h2nql -- cilium bpf ipmasq list
IP PREFIX/ADDRESS
10.0.0.0/8
169.254.0.0/16
172.16.0.0/12
192.168.0.0/16
```

The node has a direct route as: `169.254.169.254 dev eth0` 
I'm not totally sure how will Cilium eBPF react to that ?

**Comment by prune998** [comment]
I changed my ip-masq-agent configmap to be:
```
    nonMasqueradeCIDRs:
    - 10.0.0.0/8
    - 172.16.0.0/12
    - 192.168.0.0/16
    masqLinkLocal: true.  <-------------------
```

That solved the issue for pod->IMDS issue.
But Cilium pods (using HOstNetwork) are still having issues it seems...

**Comment by prune998** [comment]
Ok sorry for yesterday's noise. 
For me, the problem is down to:

The file `/etc/cni/net.d/05-cilium.conf` seems to never be created... 

Because I'm using a custom CNI config (to select eth1 as first ENI), the `postStart` script is not creating it:
https://github.com/cilium/cilium/blob/774fc6a59854feb8f51943773345ce9d2e4ab92a/plugins/cilium-cni/cni-install.sh#L84-L91

My `cilium-config` configmap is setting `write-cni-conf-when-ready: /host/etc/cni/net.d/05-cilium.conflist`
So the agent should create it when it's started.

This leaves Docker network unconfigured, so `cilium-agent` is not ready, so the node is not ready, so we have all sorts of errors in cilium-agent and cilium-operator.

What I don't know is if it's the agent waiting on the Operator or something else preventing it to initialize correctly...

**Comment by prune998** [comment]
It feels like the Cilium Agent is creating the `ciliumnode` CR, but the Operator is having a problem with it, preventing the agent to finish initialization.

back to 
```
level=info msg="Discovered new CiliumNode custom resource" name=ip-10-253-2-27.us-west-2.compute.internal subsys=ipam
level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-079c49efed379138c name=ip-10-253-2-27.us-west-2.compute.internal subsys=ipam
level=warning msg="Instance not found! Please delete corresponding ciliumnode if instance has already been deleted." instanceID=i-079c49efed379138c name=ip-10-253-2-27.us-west-2.compute.internal subsys=ipam
```

I just can't understand why it's not working but something is preventing the Operator to match the info in the `ciliumNode` with the data it finds/cache from AWS API....

**Comment by prune998** [comment]
FYI, I tested and 1.10.7 is also OK. Issue is in 1.10.8+ up to 1.11.2

**Comment by anwbtom** [comment]
@prune998 Thank you for your comments, it helped us debug issues we had a bit further, but alas we were stuck one step further in 1.10.8+ than you.  

Our situation was that we want to use different subnet(s) for our pods than our worker nodes, for some odd reason when using cilium 1.10.8+ as the CNI for EKS it won't attach an actual ENI to the node to use for pod addresses and the "Instance not found!" error was found in the cilium operators.  
When we manually attached an ENI to the node (which contained the correct subnet) the error of instance not found was gone, but in our case it stopped a bit further because it could not find the security groups used by the node (probably to assign it to the pod eni), to be clear this was not a permission denied error, since couldtrail was clean of errors.

Hope this extra information can/might help somebody else or to resolve the actual issue. :)

**Comment by prune998** [comment]
@anwbtom I can confirm the same kind of behaviour... when a node is upgraded, all the ENIs/IPs already on the node works and can be used. Attaching new ENIS or grabbing new IPs will fail. 

Now the questions is how to get someone assigned to this issue ? We need to get more traction here...

**Comment by anwbtom** [comment]
Well, in my testcase after the initial working ENI was attached it was able to attach a new ENI as well, but *only* with a SecurityGroupId pre-configured (so using the default of the node didn't work, since then this message was shown: `level=warning msg="Unable to create interface on instance: unable to get security groups failed to get security group ids" instanceID=i-0c5a3a0fa068db10f name=ip-10-253-2-252.us-west-2.compute.internal subsys=ipam`
But with (as an example) this cni-config:
`apiVersion: v1
data:
  cni-config: |
    {
      "name": "cilium-cni",
      "cniVersion": "0.3.1",
      "plugins": [{
        "cniVersion": "0.3.1",
        "type": "cilium-cni",
        "eni": {
          "subnet-tags": {
            "Target": "pod"
          },
          "vpc-id": "vpc-{my-id}",
          "first-interface-index": 1,
          "security-groups": ["sg-04bd7f5b9afd61e69"]
        }
      }]
    }`
it was able to attach a second one (key being the eni.security-groups bit)

**Comment by prune998** [comment]
```
The security group ids attached to the ENI are computed in the following order:

 1. The field ``spec.eni.security-groups`` is consulted first. If this is set
    then these will be the security group ids attached to the newly created ENI.
 2. The filed ``spec.eni.security-group-tags`` is consulted. If this is set then
    the operator will list all security groups in the account and will attach to
    the ENI the ones that match the list of tags passed.
 3. Finally if none of the above fields are set then the newly created ENI will
    inherit the security group ids of ``eth0`` of the machine.
```

It shouldn't (and wasn't) a problem before 1.10.8.
Also, a diff between 1.10.7 and 1.10.8 does not show any changes in the EKS ipam/eni code... :/

**Comment by anwbtom** [comment]
Am not saying its not a bug or that its logical (I don't know to much from the internals of cilium yet).  

But the error states its having trouble finding the security group ids: `Unable to create interface on instance: unable to get security groups failed to get security group ids`.  
So by giving it one or more ids (priority 1 according to your attached list), it won't have to search for it, and just apply them. So its more a work-around the issue than actually fixing whatever causes it.

**Comment by prune998** [comment]
Yeah, i'm testing this scenario right now and trying to add debug... I'm wondering if the issue could be a change in the AWS SDK, as it totally feels it's a mismatch between what we get from AWS, the Cilium Operator cache and what's in the `CiliumNode`

**Comment by prune998** [comment]
I added some debug in some functions and It seems to get down to:

```
func (n *Node) getSecurityGroupIDs(ctx context.Context, eniSpec eniTypes.ENISpec) ([]string, error) {
	// 1. check explicit security groups associations via checking Spec.ENI.SecurityGroups
	// 2. check if Spec.ENI.SecurityGroupTags is passed and if so filter by those
	// 3. if 1 and 2 give no results derive the security groups from eth0
``` 
If we don't set 1) or 2), eth0 is used by default.
In that case, we loop each interfaces, hoping to find the number `0`:

https://github.com/cilium/cilium/blob/79645bd18043212f3ff69b32cf7245f3644cb098/pkg/aws/eni/node.go#L276-L284

What I found, so far, is that:
- the ENIs that a looked up are only the ones created by Cilium, so `eth0` is not scanned, so this function will always "fail" (return empty security groups)
- `e, ok := rev.Resource.(*eniTypes.ENI)` always fail, meaning something is not correct with `rev.Resource`.

**Comment by prune998** [comment]
I can confirm the behaviour is different with Cilium 1.10.7, where `eth0` is part of the list of interfaces.
Now I'm investigating around the updates that lead to this behaviour change....

**Comment by prune998** [comment]
More research... 
I'm trying to get new nodes working. 
so far, when a new node is joining the cluster, `cilium agent` kicks of, create the `ciliumNode` then the `cilium-operator` tries to add ENIs/ IPs.

To do that, the Operator look in EC2 for vpcs, subnets, instances. Then this "ec2 client cache" is used to match the instance.

Because of the split network, the new EC2 instance only have one ENI (eth0) and is using the default VPC subnet (not the pod subnet).

Here's the code used to get the `instances` info :
https://github.com/cilium/cilium/blob/8a8c6ff3afc380072d54004714149c0177e65272/pkg/aws/ec2/ec2.go#L217

In the end, the instance list in the cache does not includes the new instance as the subnets does not matches, leading to `Instance not found` errors. 

Again, I still don't see why it's working with 1.10.7 as nothing changed in there...

**Comment by gandro** [comment]
So looking at the list of changes between 1.10.7 and 1.10.8, one relevant change that I see is the aforementioned https://github.com/cilium/cilium/pull/18478

@prune998 @anwbtom How are you configuring subnet tag filtering? Only using the CNI ConfigMap, or also using the `subnet-tags-filter` on `cilium-operator`? (Helm value: `eni.subnetTagsFilter="foo=true"`)

If you are using the Helm value, does it work for you if you remove it and only rely on the CNI config ConfigMap?

**Comment by anwbtom** [comment]
I was adding it on both levels e.g. in the cni-config and the eni.subnetTagsFilter.  
Will test it with only setting it on cni-config level, either today otherwise monday and come back to this.
Thanks for all the additional information and feedback!

**Comment by gandro** [comment]
Thanks. So my suspicion is that `eni.subnetTagsFilter` is broken for some reason. It likely has been broken for a while, because of bug #18328 it was actually silently ignored in `v1.10.7` and older (i.e. it didn't filter out any subnets at all). Now that we fixed  #18328 it seems like this surfaced another bug in our subnet filtering logic.

This needs to be understood and fixed, but a potential workaround is to not use `eni.subnetTagsFilter` at all and only rely on the CNI config ConfigMap.

**Comment by prune998** [comment]
OK, thanks @gandro it seems that removing the value from the `helm` values does it. 
The `instance cache` now does includes all the subnets, and the instance values are found by the operator...

**Comment by prune998** [comment]
I still don't understand why setting it twice would change the behaviour... any clue ?

**Comment by anwbtom** [comment]
It might be smart to alter the documentation with this warning in the meantime?

**Comment by gandro** [comment]
> I still don't understand why setting it twice would change the behaviour... any clue ?

The two filters work independently:
- First, `eni.subnetTagsFilter` is used as a global filter, to remove any non-matching subnets when the operator selects candidate subnets to attach ENIs from. See here: https://github.com/cilium/cilium/blob/314ca7baeff4f568ffc0bad95124e40665b1f88c/pkg/ipam/allocator/aws/aws.go#L40)
- Secondly, the CNI config section is copied over into `CiliumNode.Spec.ENI.SubnetTags` and then used to filter out candidate subnets for this particular node (where we also filter out e.g. any subnets that are not part of the node's availability zone). See here: https://github.com/cilium/cilium/blob/c492154bfe6d53870a462e50fc2d8d7dd4aaa758/pkg/aws/eni/node.go#L367

So the two filters are applied a two different times. The first one on a global level when fetching the subnets from the AWS API, and the second one on a per-node level when we select subnets that the node can attach to.

It seems that the call to the AWS API is returing nothing for some reason if the filters from `eni.subnetTagsFilter` are applied. Before v1.10.7, the `eni.subnetTagsFilter` option was effectively a no-op, so no filters were used when talking to the AWS API.

**Comment by prune998** [comment]
Thanks for the details.
I can also confirm that `eni.subnetTagsFilter` IS working for me :)
But it's not doing what I thought it would....

`eni.subnetTagsFilter` filters the `ec2 instances` to manage (or consider as possible candidates). 
In my case, because fresh instance does not have any interface using a subnet that matched the filter, no instance is selected.
I guess that's this behaviour that is bad here... and does not align with the docs:
```
| eni.subnetTagsFilter | string | `""` | Filter via tags (k=v) which will dictate which subnets are going to be used to create new ENIs |
```

My guess is that the `GetInstances()` func should not rely on subnet list gathered by `GetSubnets()` (which does use the filter)

**Comment by gandro** [comment]
@prune998  That is a very good observation, you might be on to something here. This theory matches reports of my colleagues which apparently have been using `eni.subnetTagsFilter` successfully with v1.10.8, but in their case the filter _does_ match the already attached `eth0` interface, so `GetInstances` returns something for every node in their case.

**Comment by prune998** [comment]
I guess that filtering the `instances` based the the filtered `subnets` is to reduce the amount of data to store/parse/consider ?

I'm not sure what would be the right way of doing this filtering... My guess would be that it should be based on `tags` attached to the instances, that should be = to the EKS cluster... as I guess we should only consider instances that are part of the cluster where Cilium is running... 
Maybe `eks:cluster-name` or based on the nodegroup name `eks:nodegroup-name` ? Not sure what would happen when the nodeGroup is not managed though.

The quick solution would be to ensure `GetInstances()` does not use the filtered `subnets` as a filter... I could PR for that, but maybe a little bit of thinking would'nt be bad either :)

**Comment by prune998** [comment]
side note: I have multiple EKS clusters in the same VPC/subnets... so Cilium of cluster X is actually considering all the instances of cluster Y.
Thats why I guess using the subnets / subnet filter is a bad idea here. 
Instance tags would be a better option if we want to narrow down the number of instances we're considering

**Comment by gandro** [comment]
> I guess that filtering the instances based the the filtered subnets is to reduce the amount of data to store/parse/consider ?

Yes, this is my understanding too. It seems that the global filter (Helm value `eni.subnetTagsFilter`) was indeed added as a performance optimization: https://github.com/cilium/cilium/pull/10738 

Given what you said, I think the way forward here is two-fold:

1. Let's fix `GetInstances()` to return all instances, even if they are not attached to a matching subnet (yet). This is a bug fix and can be backported to the the next v1.10.x release.
2. Let's then maybe introduce an instance tag filter as you suggested. That proposal does make a lot of sense to me. But I think that should happen in a separate PR, because I think an instance tag filter would be more like a new feature rather than a bug fix, and thus I think that change shouldn't be backported to released versions as to our [backporting criteria](https://docs.cilium.io/en/v1.11/contributing/release/backports/#backport-criteria).

**Comment by prune998** [comment]
totally agree. I'll work a PR for 1) quickly !
Thanks !

**Comment by prune998** [comment]
@gandro according to the PR https://github.com/cilium/cilium/pull/10738 it feels there's a real use-case for such an option. 
Maybe this option is just badly named or documented.
I guess we don't want to revert the use of it, or some large users will cry.

maybe we should just amend the docs to be clear about this option ?
Then, I'll focus on 2) ?

**Comment by gandro** [comment]
@prune998 Good point. I didn't study the `GetInstances` implementation before. I agree that simply removing the global subnet tag filter from `GetInstances`/`describeNetworkInterfaces` would defeat it's original purpose.

If you say you want to focus on 2), does that mean changing the `GetInstances` implementation from calling `describe-network-interaces` (with filtering by subnet tags, which causes to the bug in this issue) to calling `describe-instances` (with filtering on intstance tags instead)? 

I guess I could see that work, but it would require users for whom `GetInstances` is too expensive without any filtering to add instance filters too.

**Comment by prune998** [comment]
Yes, it will be a new parameter along `eni.subnetTagsFilter`... like `eni.instanceTagsFilter` and both options will require some documentation changes to clearly state what they do : filtering ec2 instances that are part of this specific cluster.

I have to do more testing on using un-managed nodePools in EKS before putting something together.

**Comment by anwbtom** [comment]
I'm not sure if for point 1, cilium is aware if it starting up or not. 
It indeed makes sense that the subnetfilter would not check ENI's that don't comply to the filter. However if it cannot find an ENI on the node that complies with the filter and there are still ENI spots left to be attached, wouldn't it make sense it would try to attach an ENI that does comply with the filter (at least during startup)?

**Comment by prune998** [comment]
@anwbtom this is an interresting idea.... I'll double check if it's possible and if it does not defeat the oprimization purpose of this filter.
Also, I'll double check if we can make the logs more explicit, reminding the use of the filter...

**Comment by prune998** [comment]
@gandro while preparing a PR to filter ENIs based on instances tags (instances part of the k8s cluster), I realized that most of the AWS info gathering (from the `Resync` function) just get all the possible resources, like all VPC, all subnets, etc...

I keep double-triple checking the code, but I just don't see why this would be needed.
I can feel this could have a huge impact for companies that only have one AWS account with multiple VPCs/subnets/instances/clusters...

The same thing seems to be true for Alibaba, azure... 
Maybe it's used for CiliumMesh ?

**Comment by prune998** [comment]
I linked the PR for the change. This is still early work in progress.
https://github.com/cilium/cilium/pull/19181

I just decided to add the `instance Tag ` option, and if define, it has precedence over the subnet filter when it's down to finding the available instance ENIs.

**Comment by prune998** [comment]
OK I reworked the PR https://github.com/cilium/cilium/pull/19181
It is working and I can confirm that it's doing exactly what it's meant for:

```
msg="Synchronized ENI information" 
numInstances=4 
numSecurityGroups=34 
numSubnets=9 
numVPCs=1 
subsys=eni
```

Which correspond to my 4 instances in this kubernetes cluster. Not using this (or using a filter on the subnets) will return 21 instances (all the instances in this account)

**Comment by gandro** [comment]
Quick summary for anyone stumbling upon this on v1.10/v1.11:

- If you are using the Helm value `eni.subnetTagsFilter` value, please note that it requires that each instance has an ENI with a matching subnet attached when Cilium is deployed. If in doubt, _do not use_ the `eni.subnetTagsFilter` flag, and _only_ use the subnet tag filter in the CNI config (via `cni.customConf`). The documentation has been updated accordingly.
- If you are running in a large AWS subscription, Cilium v1.12 will offer a new `eni.instanceTagsFilter` (thanks to @prune998) which you may use to reduce the number of instances queried via the AWS API.
